src/server.c

```diff
@@ -683,45 +694,22 @@ void tryResizeHashTables(int dbid) {
  *
  * The function returns 1 if some rehashing was performed, otherwise 0
  * is returned. */
-int incrementallyRehash(int dbid) {
-    /* Rehash main and expire dictionary . */
-    if (server.cluster_enabled) {
-        listNode *node, *nextNode;
-        monotime timer;
-        elapsedStart(&timer);
-        /* Our goal is to rehash as many slot specific dictionaries as we can before reaching predefined threshold,
-         * while removing those that already finished rehashing from the queue. */
-        for (dbKeyType subdict = DB_MAIN; subdict <= DB_EXPIRES; subdict++) {
-            serverLog(LL_DEBUG,"Rehashing list length: %lu", listLength(server.db[dbid].sub_dict[subdict].rehashing));
-            while ((node = listFirst(server.db[dbid].sub_dict[subdict].rehashing))) {
-                if (dictIsRehashing((dict *) listNodeValue(node))) {
-                    dictRehashMilliseconds(listNodeValue(node), INCREMENTAL_REHASHING_THRESHOLD_MS);
-                    if (elapsedMs(timer) >= INCREMENTAL_REHASHING_THRESHOLD_MS) {
-                        return 1;  /* Reached the time limit. */
-                    }
-                } else { /* It is possible that rehashing has already completed for this dictionary, simply remove it from the queue. */
-                    nextNode = listNextNode(node);
-                    listDelNode(server.db[dbid].sub_dict[subdict].rehashing, node);
-                    node = nextNode;
-                }
-            }
-        }
-        /* When cluster mode is disabled, only one dict is used for the entire DB and rehashing list isn't populated. */
-    } else {
-        /* Rehash main dict. */
-        dict *main_dict = server.db[dbid].dict[0];
-        if (dictIsRehashing(main_dict)) {
-            dictRehashMilliseconds(main_dict, INCREMENTAL_REHASHING_THRESHOLD_MS);
-            return 1; /* already used our millisecond for this loop... */
-        }
-        /* Rehash expires. */
-        dict *expires_dict = server.db[dbid].expires[0];
-        if (dictIsRehashing(expires_dict)) {
-            dictRehashMilliseconds(expires_dict, INCREMENTAL_REHASHING_THRESHOLD_MS);
-            return 1; /* already used our millisecond for this loop... */
+int incrementallyRehash(void) {
+    serverLog(LL_DEBUG,"Rehashing list length: %lu", listLength(server.rehashing));
+    if (listLength(server.rehashing) == 0) return 0;
+
+    /* Our goal is to rehash as many slot specific dictionaries as we can before reaching predefined threshold,
+     * after each dictionary completes rehashing, it removes itself from the list. */
+    listNode *node;
+    monotime timer;
+    elapsedStart(&timer);
+    while ((node = listFirst(server.rehashing))) {
+        dictRehashMilliseconds(listNodeValue(node), INCREMENTAL_REHASHING_THRESHOLD_MS);
```

Comment:
maybe we can pass the exact time left to dictRehashMilliseconds instead of INCREMENTAL_REHASHING_THRESHOLD_MS (which may be almost spent at this stage).
also, maybe we should change this function to use microseconds instead of milliseconds.

Reply:
makes sense